shared/gcs/gcs.go

```diff
@@ -19,96 +19,161 @@ limitations under the License.
+// see https://godoc.org/cloud.google.com/go/storage#Query
+func List(ctx context.Context, bucketName, prefix, delim string) []string {
+	var dirs []string
+	objsAttrs := getObjectsAttrs(ctx, bucketName, prefix, delim)
+	for _, attrs := range objsAttrs {
+		dirs = append(dirs, path.Join(attrs.Prefix, attrs.Name))
+	}
+	return dirs
+}
+
+// ListDirectChildren lists direct children paths(including files and directories).
+// Note: to avoid including unnecessary directories, prefix must end with "/".
+// e.g. if there are 2 top directories "foo" and "foobar",
+// then given prefix "foo" with list both from "foo" and "foobar"
+func ListDirectChildren(ctx context.Context, bucketName, prefix string) []string {
+	return List(ctx, bucketName, prefix, "/")
 }
 
-// GetLatestBuildNumber gets the latest build number for the specified log directory
-func GetLatestBuildNumber(ctx context.Context, logDir string, sa string) (int, error) {
-	logFilePath := logDir + latest
-	log.Printf("Using %s to get latest build number", logFilePath)
-	contents, err := ReadGcsFile(ctx, logFilePath, sa)
+// Copy file from within gcs
+func Copy(ctx context.Context, srcBucketName, srcPath, dstBucketName, dstPath string) error {
+	src := createStorageObject(srcBucketName, srcPath)
+	dst := createStorageObject(dstBucketName, dstPath)
+
+	_, err := dst.CopierFrom(src).Run(ctx)
+	return err
+}
+
+// Download file from gcs
+func Download(ctx context.Context, bucketName, srcPath, dstPath string) error {
+	handle := createStorageObject(bucketName, srcPath)
+	if _, err := handle.Attrs(ctx); nil != err {
+		return err
+	}
+
+	dst, err := os.OpenFile(dstPath, os.O_RDWR|os.O_CREATE, 0755)
 	if err != nil {
-		return 0, err
+		return err
 	}
-	latestBuild, err := strconv.Atoi(string(contents))
+	src, err := handle.NewReader(ctx)
 	if err != nil {
-		return 0, err
+		return err
+	}
+	defer src.Close()
+	if _, err = io.Copy(dst, src); nil != err {
+		return err
 	}
 
-	return latestBuild, nil
+	return nil
 }
 
-//ReadGcsFile reads the specified file using the provided service account
-func ReadGcsFile(ctx context.Context, filename string, sa string) ([]byte, error) {
-	// Create a new GCS client
-	if err := createStorageClient(ctx, sa); err != nil {
-		log.Fatalf("Failed to create GCS client: %v", err)
+// Upload file to gcs
+func Upload(ctx context.Context, bucketName, dstPath, srcPath string) error {
+	src, err := os.Open(srcPath)
+	if nil != err {
+		return err
 	}
-	o := createStorageObject(filename)
-	if _, err := o.Attrs(ctx); err != nil {
-		return []byte(fmt.Sprintf("Cannot get attributes of '%s'", filename)), err
+	dst := createStorageObject(bucketName, dstPath).NewWriter(ctx)
+	defer dst.Close()
+
+	if _, err = io.Copy(dst, src); nil != err {
+		return err
 	}
-	f, err := o.NewReader(ctx)
+
+	return nil
+}
+
+// ReadFile reads the specified file
+func ReadFile(ctx context.Context, bucketName, filePath string) ([]byte, error) {
```

Comment:
Rename to Read() for consistency as all the other methods do not use file suffix.

Reply:
done